Add oci.manifest.digest, container.image.repo_digests and make container.image.tag array

ChrsMark opened the pull request:
This PR adds ~~`container.image.digest`~~ `oci.manifest.digest`, `container.image.repo_digests` fields and make `container.image.tag` an array of strings (renamed to `container.image.tags`).
This is to cover https://github.com/open-telemetry/semantic-conventions/issues/48.

Also related to https://github.com/open-telemetry/semantic-conventions/issues/72.

More analysis can be found at https://github.com/open-telemetry/semantic-conventions/issues/48#issuecomment-1614425878.

cc: @AlexanderWert @kaiyan-sheng @mlunadia

lmolkova reviewed (changes requested):
I still think OCI terminology makes more sense here and suggest using `oci.manifest.digest` and oci.manifest.tag`.This would cover images and artifacts and would also rely on well-known and standard terminology.

ChrsMark commented:
Hey @lmolkova thanks for the feedback here!

From my perspective it would give a better user experience to provide a more generic naming. In a use case where a collector ships data to a data-store when the end user will go to search for them it would be easier to have all the related `container` fields under the `container.*` namespace. I suspect that the average user would struggle to understand what `oci.manifest.digest` stands for even if they know what they are looking for.

In addition I don't see any `tag`s specific attribute in the OCI spec: https://github.com/opencontainers/image-spec/blob/main/manifest.md. Do you see sth there that totally reflects the `container.image.tag`? If only `.digest` is reflected in OCI directly then that's an extra reason for not splitting the fields in `container.*` and `oci.*` at the same time.

To my mind we should be following what runtimes/orchestrators provide which at the same time follow the OCI under the hood. Following the OCI looks more like an implementation detail and should not be exposed to the end user to my mind since it slightly changes the scope of interest. You can also have a look at my comment at https://github.com/open-telemetry/semantic-conventions/issues/48#issuecomment-1614425878 where I analyse how the various runtimes/orchestrators define the specific fields. There I also mention that k8s is based on Container Runtime Interface (CRI) which follow the OCI spec but nowhere this detail is exposed to the end user.  Indeed the CRI reports as `RepoDigests` while in `OCI` it's just `digest` from what we can see. 

Let me know what you think :).

kaiyan-sheng commented on docs/resource/container.md:
```diff
@@ -13,7 +13,8 @@
 | `container.runtime` | string | The container runtime managing this container. | `docker`; `containerd`; `rkt` | Recommended |
 | `container.image.name` | string | Name of the image the container was built on. | `gcr.io/opentelemetry/operator` | Recommended |
-| `container.image.tag` | string | Container image tag. | `0.1` | Recommended |
+| `container.image.tag` | string[] | Container image tags. An example can be found in [Docker Image Inspect](https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageInspect). Should be only the `<tag>` section of the full name for example from `registry.example.com/my-org/my-image:<tag>`. | `[v1.27.1, 3.5.7-0]` | Recommended |
```
Thinking about `tag` vs `tags` here, we do have `container.command_args` with []string type so maybe we should change it to `tags` instead here?

ChrsMark replied:
I have no strong opinion here. If we have a strong argument for making it `tags` I'm ok, however this would require a change also in ECS.

lmolkova commented:
@ChrsMark I agree on the tag part - there is no formal tag definition in the OCI.

I don't however agree that OCI manifest digest should be recorded with `container.image.digest`.
OCI manifest is wider and more general-purpose thing than container image digest. 

It describes not just container images, but non-runnable artifacts, VM images, helm charts, or anything else.

As an owner of Azure Container Registry SDK, when I report client calls, I would not even know what users are packaging, how they intend to use it. Distribution platforms such as image/artifact registries don't know either - they just implement OCI/docker v2 APIs.

`container.image.digest` focuses on the containers and the execution side only.  Introducing it would cause the following problems:
- container/artifact registries and their client libraries will have to use `container.image` attributes for non-container things (and not images). 
- OR we'll have to introduce `vm.image`, `helm.chart`, `oci.manifest` namespaces, etc 

OCI image manifest is a standard thing defined by the spec, while container image is a vague thing. k8s docs refer to the OCI spec.

TL;DR: OCI manifest digest covers a wider set of use-cases, allows to have consistent attribute in client libraries, artifact registries, and container environments. It provides common unambiguous terminology. 

The only downside, that there is a tiny learning curve to discover OCI.  For my point of view, benefits of `oci.manifest.digest`  outweigh it.

ChrsMark commented:
Thanks for the detailed explanation @lmolkova !

To your use-case I see the point, however I would avoid using a generic naming like `oci.image.digest` to report a Container's Image Digest specifically. Based on your example the `oci.image.digest` might be populated for other entities not just containers so I think it would better to use a specific field for the container entity. I would prefer using `container.image.digest` and then `registry.image.digest` to refer to Registry Images in general. I would see value in a query like `container.image.digest==registry.image.digest` since the second field creates a super set that includes the first one.   

Also, I made some extra research to try and get things together and I think that for the `container.image.digest` we need to clarify some things along with the `container.image.id`. Let me try to summarize my findings: 

So the image ID is the equivalent of the `.config.digest` from
https://github.com/opencontainers/image-spec/blob/main/manifest.md
We already depict this at `container.image.id` since https://github.com/open-telemetry/semantic-conventions/pull/39/files#r1202644093.

The current PR intends to add the Digest information that is capable to be used for downloading an image.
In Docker and the CRI it is called `RepoDigest` and it is not part of the Image Manifest at https://github.com/opencontainers/image-spec/blob/main/manifest.md.
Indeed it seems to be a different field which is part of the [Manifest List](https://docs.docker.com/registry/spec/manifest-v2-2/#manifest-list) or the "fat manifest" as it is called (instead of the [Image Manifest](https://docs.docker.com/registry/spec/manifest-v2-2/#example-image-manifest)). But still the ID and the Digest are depicted in different parts of the spec.

Example:
```console=
➜  ~ docker pull prom/prometheus:v2.16.0@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a
docker.io/prom/prometheus@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a: Pulling from prom/prometheus
Digest: sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a
Status: Image is up to date for prom/prometheus@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a
docker.io/prom/prometheus:v2.16.0@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a
➜  ~ docker manifest inspect --verbose prom/prometheus:v2.16.0 | jq '.[0].Descriptor'
{
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "digest": "sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a",
  "size": 2824,
  "platform": {
    "architecture": "amd64",
    "os": "linux"
  }
}

➜  ~ docker inspect prom/prometheus:v2.16.0 --format 'Id: {{.Id}}                                                                            
Repo Digest: {{index .RepoDigests }}' 
Id: sha256:e935122ab143a64d92ed1fbb27d030cf6e2f0258207be1baf1b509c466aeeb42                                                                            
Repo Digests: [prom/prometheus@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a prom/prometheus@sha256:e4ca62c0d62f3e886e684806dfe9d4e0cda60d54986898173c1083856cfda0f4]

➜  ~ docker manifest inspect --verbose prom/prometheus:v2.16.0 | jq '.[0].SchemaV2Manifest.config'                                           
{
  "mediaType": "application/vnd.docker.container.image.v1+json",
  "size": 6669,
  "digest": "sha256:e935122ab143a64d92ed1fbb27d030cf6e2f0258207be1baf1b509c466aeeb42"
}

```

So to summarize it, we already have the `container.image.id` which is the equivalent of `Image ID` in Docker, Kubernetes and CRI as it is also mentioned at https://github.com/open-telemetry/semantic-conventions/blob/main/model/resource/container.yaml#L35-L47. In terms of OCI it is the `digest` of the `.config` section of the manifest.

So here we talk about adding the `Repo Digest` identifier. Thus, if my claims above are correct we already skipped using the `oci.` specific namespace and to my mind it would be better if we continue doing this. It's easier to just use a mapping more friendly to the Container runtimes' (like ImageID, RepoDigest) users rather than going deep into the OCI spec which took me significant time to analyse and correlate with the fields provided by the runtimes. Plus the fact that the intend to use `oci` would extend the scope.

I would propose sth like `container.image.digest` and `registry.image.digest` (for registry use cases) as I already mentioned.

@lmolkova out of curiosity which part of the OCI spec would you be willing to use in your use-case specifically? The one that is depicted as `Repo Digest` and is used for downloading an image or the `Image ID` one?

Also I think having input from more people here to get more opinions would help a lot :).

lmolkova commented:
> To your use-case I see the point, however I would avoid using a generic naming like oci.image.digest to report a Container's Image Digest specifically. Based on your example the oci.image.digest might be populated for other entities not just containers so I think it would better to use a specific field for the container entity.

@ChrsMark , this is where we disagree. Attribute-based correlation is one of the features OTel provides. If we give the same thing multiple names we would not be able to correlate using attributes.

E.g. when building an artifact, I want to:
- record data how the artifact is obtained regardless of container environment
- record how it's pushed to the registry on the client side
- record how it's pushed on the registry side
- record how it's pulled on both sides
- record how it's used (e.g. as a resource attribute on an application)

Assuming a bright future where I can get access to all this telemetry, using `oci.manifest.digest`  I can find everything that has ever happened to this specific artifact digest. 

Querying gets more complicated with `container.image.digest`, and `registry.artifact.digest`, now you need to know they are both available and are the same thing. 

Once you add config digest and layers digest into the picture, the bigger the need for unambiguous and externally defined image id becomes (which is defined in OCI and manifest digest).

I'm still struggling to understand the problem with `oci.manifest.digest` except a small learning curve. Can you explain if there are other reasons to make attribute-based correlation more complicated and use ambiguous and vague definitions while we have a spec that defines id?

ChrsMark commented:
> I'm still struggling to understand the problem with `oci.manifest.digest` except a small learning curve. Can you explain if there are other reasons to make attribute-based correlation more complicated and use ambiguous and vague definitions while we have a spec that defines id?

Thank's @lmolkova! Let me try to collect my concerns bellow: 

1) The learning curve is not quite small, see my explanation at https://github.com/open-telemetry/semantic-conventions/pull/159#issuecomment-1641674353 of how the OCI information is depicted in runtimes. Also there was a struggle to clearly define those at https://github.com/open-telemetry/semantic-conventions/pull/39/files#r1202644093. And still the schema needs to be validated again, see point 4 bellow. All these make me feel that the learning curve would not be negligible.
2) Coming from an Infrastructure Observability background I would assume that developers crafting Observability tools/collectors would find it more straight forward to use a terminology more relative to container runtimes (`ImageDigest`, `ImageID`). Otherwise they also need to go through the learning curve.
3) If we follow the `oci.image.digest` we might hit an issue with this field containing a super sets of objects/entities: A container image is an OCI image but an OCI image is not necessarily a container image. So if I plot in a pie chart the top 10 OCI images this might be misleading if I don't know that this dataset does not only contain container image information. This is an example illustrating the points 1 and 2. It's a clear issue of being generic VS being specific.
4) Last but not least, what is the plan about the `container.image.id` field that we have already introduced with https://github.com/open-telemetry/semantic-conventions/pull/39/files#r1202644093? As I explained at https://github.com/open-telemetry/semantic-conventions/pull/159#issuecomment-1641674353 we would need to validate it again since this is also part of the OCI spec. @joaopgrassi @marcsanmi what are your thoughts on this?

Still open question:
A) which part of the OCI spec would you be willing to use in your use-case specifically? The one that is depicted as Repo Digest and is used for downloading an image or the Image ID one? 

I'm not totally against the `oci.*` proposal but I would like to address the above first.

lmolkova commented:
@ChrsMark 

Thank you for the update!

1. I'm not sure I understand the concern, please take a look at p4 below
2. My take (coming from an observability perspective) is that using standard, unambiguous terminology and being able to correlate telemetry using attributes is quite important. 
3. It probably means that the pie chart needs to take other information into account - such as manifest media type. BTW, for the registry scenario, it does not matter what type manifest has since the registry pulls and pushes arbitrary data and does not care.
4. The  `container.image.id`, as we discussed before is a container-runtime-specific thing that represents the image id - in case of docker it's (as you figured out, the config digest), in the case of k8s it's something else. The same image has different ids on docker and k8s.
If there are other environments `container.image.id` would be different in each of them,  `oci.manifest.digest` would be a cross-environment, unified digest, common across different environments.
I.e. in your Prometheus example above,

`container.image.id=sha256:e935122ab143a64d92ed1fbb27d030cf6e2f0258207be1baf1b509c466aeeb42`
`oci.manifest.digest=sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a`.

I don't see how calling it `container.image.digest` would be any easier to understand or use (on the contrary, we'll have yet another attribute to correlate it to registry telemetry). Plus if I google for `container.image.digest` I will get nothing definitive, `oci.manifest.digest` will bring me right to the spec.  


> A) which part of the OCI spec would you be willing to use in your use-case specifically? The one that is depicted as Repo Digest and is used for downloading an image or the Image ID one?

If I defined oci namespace for my SDK, I'd start with:
- `oci.manifest.digest`
- `oci.manifest.media_type`  
- `oci.schema_version`  (2 by default)

thinking more about it, I'd be useful for me to record config and layers digests on telemetry, so I'd also consider
- `oci.blob.digest` (blob is coming from docker v2 API, but )
- `oci.blob.media_type` - this can be used to distinguish configs from layers
- `oci.schema_version`

Since SDK (or registry) has no knowledge of container environment the image/artifact will be used in (or if  it will be used in the container environment at all), it would not know anything about containers or their ids.

ChrsMark commented:
Hey @lmolkova and thank you for the feedback! I see how `container.image.id` is different and now things are clear to me.
I think we can decide on using the `oci.manifest.digest`, I will update this PR accordingly :).

jsuereth commented on docs/resource/oci.md:
```diff
@@ -0,0 +1,15 @@
+# OCI
```
Should this be a subsection under `container.md`?  I.e. OCI specifics?

AlexanderWert replied:
+1 on moving this markdown section under `container.md`. 
Actually, another good example in which [decoupling attribute definition from usage would help](https://github.com/open-telemetry/semantic-conventions/issues/197).

lmolkova reviewed (approved):
LGTM, and thank you for the great discussion, @ChrsMark !

One small comment from me on the  `container.image.id` - the note on it seems out of date - it says "OCI defines a digest of manifest ." (I can't leave a comment on it)

I'd either remove this sentence completely or change it to something along the following lines:

"The `container.image.id` of the same image running in different environments don't not always match. The `oci.manifest.digest` attribute, however, is the same for a given image in all container runtimes that follow OCI specification."

ChrsMark commented:
Thanks for reviewing this folks! In adacd83 , I changed the fields to plural form since I think it's more accurate based on https://opentelemetry.io/docs/specs/otel/common/attribute-naming/#name-pluralization-guidelines. Runtimes also report those in plural form. 

I also tuned the descriptions accordingly.
@lmolkova let me know what you think about the `oci.manifest.digests` description specifically.

joaopgrassi commented on model/resource/oci.yaml:
```diff
@@ -0,0 +1,22 @@
+groups:
+  - id: oci
```
```suggestion
  - id: oci.manifest
```

joaopgrassi commented on model/resource/oci.yaml:
```diff
@@ -0,0 +1,22 @@
+groups:
+  - id: oci
+    prefix: oci
```
```suggestion
    prefix: oci.manifest
```
I think it makes sense to define the manifest as a "top-level" since we may use other OCI concepts later on? Also it makes defining the attributes below shorter.

joaopgrassi commented on model/resource/oci.yaml:
```diff
@@ -0,0 +1,22 @@
+    brief: >
+      An OCI instance.
+    attributes:
+      - id: manifest.digests
```
```suggestion
      - id: digest
```
The property is `digest` singular, right? https://github.com/opencontainers/image-spec/blob/main/descriptor.md#digests

> The value of the `digest` property is a string consisting of an algorithm portion and an encoded portion

joaopgrassi commented on model/resource/oci.yaml:
```diff
@@ -0,0 +1,22 @@
+      - id: manifest.digests
+        type: string
+        brief: >
+          The digest(s) of the OCI image manifest. For container images specifically it can be one or more
```
Do we have more than one "digest"? If it is more, shouldn't this be an array then? The name/type is a bit contradicting right now.

ChrsMark replied:
Good catch, thanks!

The intention is to be plural and hence it should be an array. Fixed the type with [9e3d8fb](https://github.com/open-telemetry/semantic-conventions/pull/159/commits/9e3d8fb4ada03898a25f03085a3495cf2d9535b8).

From container runtimes we can get more then one digests related to a specific running image. Hence we need to have it as an array to report multiple values.

From the repository perspective we can either report only the basic digest or if more digests are available (from additional image layers maybe) report all of them as well. @lmolkova would that make sense?

joaopgrassi replied:
Makes sense! Feel free to close this.

lmolkova replied:
sorry, I did not notice this  - the image has exactly one manifest (which can itself refer to a list of manifests)

joaopgrassi replied:
I think this is resolved now. Closing.

joaopgrassi commented on docs/resource/container.md:
```diff
@@ -13,17 +13,31 @@
 **[2]:** If using embedded credentials or sensitive data, it is recommended to remove them to prevent potential leakage.
 <!-- endsemconv -->
 
+## OCI specifics
```
```suggestion
## Open Container Initiative (OCI)
```

Maybe here we can add some very brief sentence about OCI and leave a link to their spec? 

And I suggest we add these conventions under their own section, such as:

```
### OCI Image Manifest
```

joaopgrassi commented on docs/resource/container.md:
```diff
@@ -13,17 +13,31 @@
+
+**type:** `oci`
+
+**Description:** An OCI instance.
```
Is it really an instance? Maybe something like: Describes a manifest that can be used by... then maybe grab some of the great research you folks did in the discussions?

joaopgrassi commented on model/resource/container.yaml:
```diff
@@ -44,7 +47,9 @@ groups:
           /namespace/service/dockerfile@sha256:bdeabd40c3a8a492eaf9e8e44d0ebbb84bac7ee25ac0cf8a7159d25f62555625"`.
 
-          OCI defines a digest of manifest.
+          The ID of the same image running in different environments don't not always match.
```
```suggestion
          The ID of the same image running in different environments do not always match.
```

joaopgrassi commented on model/resource/container.yaml:
```diff
@@ -44,7 +47,9 @@ groups:
-          OCI defines a digest of manifest.
+          The ID of the same image running in different environments don't not always match.
+          The `oci.manifest.digests` attribute, however, is the same for a given image
+          in all container runtimes that follow OCI specification.
```
Was the goal of this sentence to say something like: If you want to have "something" that no matter where it runs will be always the same, use `oci.manifest.digests`? 

If so, then maybe we can tell it in something like this:?

> Consider using `oci.manifest.digests` if it is important to identify the same image in different environments/runtimes

joaopgrassi commented:
Sorry @ChrsMark for not responding earlier. I read all the conversation above and I'm glad we reached an agreement. I agree with the points @lmolkova brought up, I think adopting OCI is more "future proof" and gives us the power of correlating things. I left a review with a few points.

joaopgrassi reviewed (approved):
I think we only need an entry on the changelog now!

AlexanderWert commented:
@open-telemetry/specs-semconv-maintainers I think this can be merged?

AlexanderWert commented:
@joaopgrassi Chris is out this week. I just resolved the conflict in the Changelog file.

lmolkova commented on model/resource/oci.yaml:
```diff
@@ -0,0 +1,24 @@
+    brief: >
+      An OCI image manifest.
+    attributes:
+      - id: digests
+        type: string[]
```
```suggestion
      - id: digest
        type: string
```


Image/artifact has one manifest (which can have multiple nested manifests or layers)

lmolkova reviewed (changes requested):
I just noticed one small issue and then we should be good to go

ChrsMark commented:
@lmolkova as I explained at https://github.com/open-telemetry/semantic-conventions/pull/159#discussion_r1283357024, container runtimes like Docker and CRI can provide multiple Repo Digests for a container image, see for example https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageInspect.

Based on this I think we should give the option to populate this field with multiple values as well. Otherwise if we make it singular/`string`, it will not be able to cover many important use cases. If this does not make sense from "oci" perspective then maybe we should reconsider the decision of using the "oci" generic namespace and maybe think of introducing 2 different fields `oci.manifest.digest` and `container.image.digests`(maybe that's more accurate based on this last observation/evaluation?).

Both options would be ok from my side, so let me know what you think @joaopgrassi @lmolkova :).

lmolkova commented:
> @lmolkova as I explained at [#159 (comment)](https://github.com/open-telemetry/semantic-conventions/pull/159#discussion_r1283357024), container runtimes like Docker and CRI can provide multiple Repo Digests for a container image, see for example https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageInspect.
> 
> Based on this I think we should give the option to populate this field with multiple values as well. Otherwise if we make it singular/`string`, it will not be able to cover many important use cases. If this does not make sense from "oci" perspective then maybe we should reconsider the decision of using the "oci" generic namespace and maybe think of introducing 2 different fields `oci.manifest.digest` and `container.image.digests`(maybe that's more accurate based on this last observation/evaluation?).
> 
> Both options would be ok from my side, so let me know what you think @joaopgrassi @lmolkova :).

The `oci.manifest.digest` does not capture the `RepoDigests` and there could be just one for the running container.
`RepoDigest` is docker-specific thing. E.g. k8s uses the repo-scoped-digest in the imageID (e.g.  `mcr.microsoft.com/oss/busybox/busybox@sha256:582a641242b49809af3a1a522f9aae8c3f047d1c6ca1dd9d8cdabd349e45b1a9`)

And the repo-digest (whatever they are) would be already populated in the `container.image.id` attribute.

If we need to capture docker-only RepoDigests, let's do it in the `container.docker.image.digests` or just `docker.image.digests` since other runtimes don't share the same property.

ChrsMark commented:
Hey @lmolkova , let me try to clarify what I have in mind :) :


 > The `oci.manifest.digest` does not capture the `RepoDigests` and there could be just one manifest for the container instance. The digest has specific format (hash_algo:hash), so it cannot contain the `internal.registry.example.com:5000/example@` part.
> 

From the docs' page [example]( https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageInspect) we are not going to add the `internal.registry.example.com:5000/example@` part but the rest of it. 

From a previously used example:

```console
➜  ~ docker pull prom/prometheus:v2.16.0@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a
docker.io/prom/prometheus@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a: Pulling from prom/prometheus
Digest: sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a
Status: Image is up to date for prom/prometheus@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a
docker.io/prom/prometheus:v2.16.0@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a


➜  ~ docker inspect prom/prometheus:v2.16.0 --format 'Id: {{.Id}}                                                                            
Repo Digest: {{index .RepoDigests }}' 
Id: sha256:e935122ab143a64d92ed1fbb27d030cf6e2f0258207be1baf1b509c466aeeb42                                                                            
Repo Digests: [prom/prometheus@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a prom/prometheus@sha256:e4ca62c0d62f3e886e684806dfe9d4e0cda60d54986898173c1083856cfda0f4]
```

Both `docker pull prom/prometheus@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a` and `docker pull prom/prometheus@sha256:e4ca62c0d62f3e886e684806dfe9d4e0cda60d54986898173c1083856cfda0f4` will work, so those are valid digests.

So in the bare minimum example of docker developers that want to capture the `digests` would need to iterate over the `RepoDigests` field array and extract the digests' part.

> `RepoDigest` is docker-specific thing. E.g. k8s uses the repo-scoped-digest as the imageID (e.g. `mcr.microsoft.com/oss/busybox/busybox@sha256:582a641242b49809af3a1a522f9aae8c3f047d1c6ca1dd9d8cdabd349e45b1a9`)
> 
> And the repo-digest (whatever it is) would be already populated in the `container.image.id` attribute.
> 

For k8s, I don't think that is the case. Kubernetes Container objects will not hold that information anyways. The runtime, or better the [CRI](https://kubernetes.io/docs/concepts/architecture/cri/), will provide the `RepoDigests`.
For example if you `exec`/`ssh` inside a k8s node you can `crictl` to get the container images (for example with `containerd` as runtime).

An example:

```json
➜  ~ crictl image -o json
 {
      "id": "sha256:e935122ab143a64d92ed1fbb27d030cf6e2f0258207be1baf1b509c466aeeb42",
      "repoTags": [
      ],
      "repoDigests": [
        "docker.io/prom/prometheus@sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a"
      ],
      "size": "54439175",
      "uid": null,
      "username": "nobody",
      "spec": null,
      "pinned": false
    },
```

From this it's clear to me that CRI provides that info in plural form (not just for docker but in general) and we should be aligned with that because that's what the APIs provide. See how the API's proto is defined: https://github.com/kubernetes/cri-api/blob/c75ef5b473bbe2d0a4fc92f82235efd665ea8e9f/pkg/apis/runtime/v1/api.proto#L1238.
In the above example the `sha256:efd99a6be65885c07c559679a0df4ec709604bcdd8cd83f0d00a1a683b28fb6a` digest can be used to `pull` the specific image from the registry, so it's an actual digest. 

So again an easy way to extract the digests in k8s containers is to access the CRI API and use the `repoDigests` field array in order to extract the digests.

> If we need to capture docker-only RepoDigests, let's do it in the `container.docker.image.digests` or just `docker.image.digests` since other runtimes don't share the same property.

Based on the above examples it should be `container.image.digests` since it's also aligned with the CRI not just docker.

joaopgrassi commented:
I'm far from being an expert in this subject, but I read all again and I see points in both sides. 

Reading again the [OCI image spec](https://github.com/opencontainers/image-spec/blob/main/manifest.md#image-manifest-property-descriptions) it has a property called `config` which holds the [descriptor](https://github.com/opencontainers/image-spec/blob/main/descriptor.md#properties). In the descriptor docs it's clear that it talks about a single "targeted content", and that has a single `digest` - the spec even says:

>A Content Descriptor includes the type of the content, **a content identifier (digest)**, and the byte-size of the raw content. Optionally, it includes the type of artifact it is describing.

It seems if we take OCI for it, makes sense to use what they also say there - a single `digest`. But with the cases that @ChrsMark brought, I'm not sure. Maybe we need to think about how instrumentations would populate this field, so we can have more info to base the decision here? 

If we don't know and want to be flexible, maybe going with the array approach might be a good trade off? Not sure..

ChrsMark commented:
Thanks for the feedback @joaopgrassi!

I agree that based on [OCI image spec](https://github.com/opencontainers/image-spec/blob/main/manifest.md#image-manifest-property-descriptions) the manifest's digest should be only one and in singular form.

My intention to make it plural/array was based on the the need to support the Runtimes' use cases (see https://github.com/open-telemetry/semantic-conventions/pull/159#issuecomment-1685786852). I see that it is a compromise and less accurate.


> Maybe we need to think about how instrumentations would populate this field, so we can have more info to base the decision here?

To illustrate my point I'm mostly interested to cover use cases of infra/platform observability/security: As an analyst I want to  look for a specific digest if exists in the container images that are present in my cluster. This should be equivalent to fetching everything from the CRI API manually and analyzing manually the information.


Reflecting all the above conversations I have the feeling that in order to be accurate we need to be specific and avoid any generalization at this point. 
From my point of view the most accurate is to introduce 2 different fields:
1) `oci.manifest.digest` to cover all the generic usecases as described at https://github.com/open-telemetry/semantic-conventions/pull/159#issuecomment-1651077018
2) `container.image.digests` to cover information retrieved from the CRI APIs as described at https://github.com/open-telemetry/semantic-conventions/pull/159#issuecomment-1685786852 to specifically reflect what the CRI API provides.

Let me know what you folks think.

joaopgrassi commented:
Sounds reasonable to me, but we would have duplication of data right? As container.image.digests would be a superset of `oci.manifest.digest`. In cases where @lmolkova described where it deals with "non container" scenarios we would only populate `oci.manifest.digest` if I got it right?

ChrsMark commented:
I think it depends on the perspective/method we collect these data:

>  but we would have duplication of data right? As container.image.digests would be a superset of oci.manifest.digest.
1) In cases we only have access to the already stored container images of our observed system we can retrieve the information from the CRI's API and hence we would populate `container.image.digests` accordingly. We don't have a straight forward way to retrieve the single manifest's digest so far. In some cases we might be able to populate both fields but that's a fair compromise for now. Maybe in the long run the CRI definition will change (to singular) and we can deprecate the `digests` in order to only use the oci one.

> In cases where @lmolkova described where it deals with "non container" scenarios we would only populate oci.manifest.digest
2) In cases we have direct access to know the specific manifest's digest we will be using the `oci.manifest.digest` (for example in download time/execution).

lmolkova commented:
There is 1:1 relationship between container and an image. One container can run only one image. 

There is also 1:1 relationship between an image and it's manifest. 

So having multiple digests for one container does not make sense.

The smart image can be pushed to multiple repositories, but if it's the same image, it will have the same oci digest anywhere (as it's a sha256 of manifest). Check out answers on this thread: https://stackoverflow.com/questions/45533005/why-digests-are-different-depend-on-registry 


If we need to support docker v1 or something else where the same image can have multiple manifest digests, let's create `contaiee.docker.repp_digests` or something similar.

ChrsMark commented:
All right, based on the above discussions I have changed the `oci.manifest.` field accordingly to singular and introduced a new one to reflect what CRI's api provides:

1) `oci.manifest.digest`
2) `container.image.repo_digests` to reflect what CRI provides (https://github.com/kubernetes/cri-api/blob/c75ef5b473bbe2d0a4fc92f82235efd665ea8e9f/pkg/apis/runtime/v1/api.proto#L1238) for infra observability and security use-cases.

I hope that covers all that we have discussed so far.

ChrsMark commented:
@lmolkova I think the [requested changes](https://github.com/open-telemetry/semantic-conventions/pull/159#pullrequestreview-1572278760) are now covered :) . Are we good to go with this one?

lmolkova commented on model/resource/container.yaml:
```diff
@@ -44,8 +47,19 @@ groups:
+          [Docker](https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageInspect) and
+          [CRI](https://github.com/kubernetes/cri-api/blob/c75ef5b473bbe2d0a4fc92f82235efd665ea8e9f/pkg/apis/runtime/v1/api.proto#L1237-L1238)
+          report those under the `RepoDigests` field.
+        examples: [ 'sha256:e4ca62c0d62f3e886e684806dfe9d4e0cda60d54986898173c1083856cfda0f4', 'sha256:3c3a4604a545cdc127456d94e421cd355bca5b528f4a9c1905b15da2eb4a4c6b' ]
```
nit: I think it should be full ones that include repo URL
```suggestion
        examples: [ 'example@sha256:e4ca62c0d62f3e886e684806dfe9d4e0cda60d54986898173c1083856cfda0f4', 'internal.registry.example.com:5000/example@sha256:3c3a4604a545cdc127456d94e421cd355bca5b528f4a9c1905b15da2eb4a4c6b' ]
```

ChrsMark replied:
Good catch! Fixed at e5a3293, thank you!

lmolkova reviewed (approved):
Left one small comment, otherwise LGTM. Thank you!

ChrsMark commented:
@open-telemetry/specs-semconv-maintainers this one should be ready to for merge?
